### ISSUE TITLE
Clarify TableOrigin interface and scope

### ISSUE THREAD

**Issue by jfcorbett** [startable/pdtable#35]
Right now we have a `TableOriginCSV` object. 
Should also have one for Excel files, perhaps for streams as well?
To what extent can they be combined?

**Comment by jfcorbett** [comment]
@JanusWesenberg Comments on this issue please?  This is my next clean-up project! It would be nice to be aligned up front.

**Comment by JanusWesenberg** [comment]
Yes, that part was at best sketched. My thinking was something along the following.

The table origin data must somehow reference metadata for the file being read. This metadata must be passed into the reader by the caller, and could be a file path relative to a project root, records system metadata, a blob key or any other file-level metadata. Based on this, the reader must then be able to create a table-level origin object suitable for the caller.

One way of doing this could be to provide a `TableOriginPolicy` object as input `origin` (or maybe rename that input).
```

class TableOriginPolicy(Protocol):
    @abstractmethod
    def origin_for(self, **kwargs) -> TableOrigin:
        """
        Provide table origin for the location specified by keyword arguments

        Table sources are free to call this method with any arguments, but should
        prefer the convenience methods `origin_for_line` and `origin_for_workbook_location`.
        """
        pass

    def origin_for_line(self, line: int) -> TableOrigin:
        """
        Returns origin for line in text-based input source such as CSV.
        """
        return self.origin_for(line=line)

    def origin_for_workbook_location(self, row: int, column: str, sheet: Optional[str] = None) -> TableOrigin:
        """
        Returns origin for line in text-based input source such as CSV.
        """
        return self.origin_for(row=row, column=column, sheet=sheet)
```

In terms of actually building a full reader system, we will need a broader context setup -- as we discussed earlier, the pipeline processing as currently sketched will not be able to handle includes etc. Whether this should be part of the core I am not sure.
